test/expect-tests/dune_rpc_e2e/dune_rpc_e2e.ml

```diff
@@ -230,22 +230,22 @@ let setup_diagnostics f =
           | Remove e -> Remove (sanitize e)
         in
         printfn "%s" (Dyn.to_string (Dune_rpc.Diagnostic.Event.to_dyn e))
-      in
-      Client.Handler.create
-        ~diagnostic:(fun de ->
-          List.iter de ~f:on_diagnostic_event;
-          Fiber.return ())
-        ()
     in
-    run_client ~handler (fun client ->
+    run_client (fun client ->
         (* First we test for regular errors *)
         files [ ("dune-project", "(lang dune 3.0)") ];
-        let* () =
-          printfn "subscribing to notifications";
-          Client.notification client Dune_rpc.Public.Notification.subscribe
-            Diagnostics
+        let* sub, stream =
+          printfn "subscribing to diagnostics";
+          Client.subscribe client Dune_rpc.Sub.diagnostic
         in
-        f client)
+        Fiber.fork_and_join_unit
+          (fun () ->
+            let* () = f client in
+            Client.Subscription.cancel sub)
```

Comment:
This looks like it's going to race against `sequential_iter`. Is there any mitigation?